### PR TITLE
Hookify Dropzone

### PR DESCRIPTION
**Pull request by AndrewMusgrave**
### WHAT is this pull request doing?

* converting dropzone to a functional component
* reduce complexity 

### How to 🎩

This is a pretty big changeset. It'll need some good tophatting. I've found a few bugs already and address them already, but just to be safe try and break this ⚒ . The readmes example are a good place to start.

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

### CODE REVIEW

File: src/components/DropZone/DropZone.tsx [hunk 1]
```diff
@@ -111,460 +96,325 @@ export interface DropZoneProps {
...
+  useEventListener(dropNode, 'dragenter', handleDragEnter);
+  useEventListener(dropNode, 'dragleave', handleDragLeave);
+  useEventListener(window, 'resize', adjustSize);
+
+  useComponentDidMount(() => {
+    adjustSize();
+    openFileDialog && triggerFileDialog();
+  });
+
+  const currentID = useUniqueId('DropZone', id);
```
Comment (BPScott) [inline src/components/DropZone/DropZone.tsx]: We solved this slightly differently in [Checkbox](https://github.com/Shopify/polaris-react/blob/cd2cb606d3b048f74b7f1ae212dbfbe2702e8f9e/src/components/OptionList/components/Checkbox/Checkbox.tsx#L21) in a way that avoids you having to do the currentID/id renaming dance

**Comment by LauraAubin** [comment]
This appears to already exist, but the example `Drop zone with a label` doesn't work, or doesn't appear to be since it's not showing my uploaded files. Was there a reason for this?

**Comment by AndrewMusgrave** [comment]
@LauraAubin My guess is to keep the example small and stateless

File: UNRELEASED.md [hunk 2]
```diff
@@ -7,6 +7,7 @@
 - Added `external` prop to `ResourceList` ([#2408](https://github.com/Shopify/polaris-react/pull/2408))
 - Added `onMouseEnter` and `onTouchStart` props to `Button` ([#2409](https://github.com/Shopify/polaris-react/pull/2409))
 - Added `ariaHaspopup` prop to `Popover` ([#2248](https://github.com/Shopify/polaris-react/pull/2248))
+- Migrated `Dropzone` to a functional component and reduce complexity ([#2360](https://github.com/Shopify/polaris-react/pull/2360))
```
Comment (dleroux) [inline UNRELEASED.md]: We usually put these under ` Code quality`

**Comment by BPScott** [comment]
Removing my request for review on this, as I'm stupid-busy with other stuff at the moment before my holiday next week so I won't have time to look at this for a while

File: src/components/DropZone/tests/DropZone.test.tsx [hunk 3]
```diff
@@ -71,7 +72,9 @@ describe('<DropZone />', () => {
   it('calls the onDrop callback with files when a drop event is fired on document', () => {
     mountWithAppProvider(<DropZone dropOnPage onDrop={spy} />);
     const event = createEvent('drop', files);
-    document.dispatchEvent(event);
+    act(() => {
+      document.dispatchEvent(event);
+    });
```
Comment (LauraAubin) [inline src/components/DropZone/tests/DropZone.test.tsx]: I'm curious why this is needed now?

File: UNRELEASED.md (hunk 2 above)
Comment (AndrewMusgrave) [reply UNRELEASED.md]: However, in this situation, it's not just a migration. I've changed the business logic. Even so, do you think it should still be in `code quality`? I could lean towards `Enhancements ` or `Enhancements `, thoughts?

Comment (dleroux) [reply UNRELEASED.md]: It works here, the key being "and reduce complexity".

File: src/components/DropZone/tests/DropZone.test.tsx (hunk 3 above)
Comment (dleroux) [reply src/components/DropZone/tests/DropZone.test.tsx]: [ ...that makes sure all updates related to these “units” have been processed and applied to the DOM before you make any assertions...](https://reactjs.org/docs/testing-recipes.html#act)

You'll get an error in the console if you didn't have it.

**Pull request merged**